src/server/main_service.cc

```diff
@@ -1038,6 +1038,62 @@ void Service::Function(CmdArgList args, ConnectionContext* cntx) {
   return (*cntx)->SendError(err, kSyntaxErrType);
 }
 
+void Service::PubsubChannels(string_view pattern, ConnectionContext* cntx) {
+  vector<vector<string>> result_set(shard_set->size());
+
+  shard_set->RunBriefInParallel([&](EngineShard* shard) {
+    result_set[shard->shard_id()] = shard->channel_slice().ListChannels(pattern);
+  });
+
+  vector<string> union_set;
+  for (auto&& v : result_set) {
+    union_set.insert(union_set.end(), v.begin(), v.end());
+  }
+
+  (*cntx)->SendStringArr(union_set);
+}
+
+void Service::PubsubPatterns(ConnectionContext* cntx) {
+  (*cntx)->SendLong(
```

Comment:
nit: pls separate. introduce a variable `size_t pattern_count = ...` and then SendLong